src/sdk/service/network-validation-service/network-validation-service.ts

```diff
@@ -0,0 +1,468 @@
+  public static getSubnetIPv4AddressRangeStrWithoutReserved(netmask: string, ipAddress: string): string | null {
+    const range = this.getSubnetIPv4AddressRange(netmask, ipAddress);
+    if (!range) {
+      return null;
+    }
+    const modIPv4AddressRange = {
+      start: this.getIntegerAsIPv4Address(this.getIPv4AsInteger(range.start) + 1),
+      end: this.getIntegerAsIPv4Address(this.getIPv4AsInteger(range.end) + 1)
+    };
+    return modIPv4AddressRange.start + '-' + modIPv4AddressRange.end;
+  }
+
+  /**
+   * Indicates whether an IP address belongs to a subnet. Returns null if any of
+   * the parameters do not represent valid network identifiers.
+   * @param {string} netmask: the netmask of the subnet.
+   * @param {string} netIpAddress: any IP address on the subnet.
+   * @param {string} testIpAddress: the IP address which should be tested for
+   *          whether it belongs to the subnet.
+   * @return {boolean} boolean value.
+   */
+  public static isInSubnet(netmask: string, netIpAddress: string, testIpAddress: string): boolean | null {
+    if (!(this.isValidIPv4Address(netIpAddress) && this.isValidNetmask(netmask)
+      && this.isValidIPv4Address(testIpAddress))) {
+      return null;
+    }
+    const netmaskInt = this.getIPv4AsInteger(netmask);
+    const netIpInt = this.getIPv4AsInteger(netIpAddress);
+    const testIpInt = this.getIPv4AsInteger(testIpAddress);
+    return ((netmaskInt & netIpInt) === (netmaskInt & testIpInt));
+  }
+
+  /**
+   * Indicates whether an IP address is in a range of IP addresses. Returns null
+   * if any of the parameters do not represent valid network identifiers.
+   * @param {string} ipAddress: the IP address which should be tested for whether
+   *          it is in the range.
+   * @param {IPv4AddressRange} range: an object of the form { start: "x.x.x.x", end:
+   *          "x.x.x.x" }, representing the range of IP addresses.
+   * @return {boolean} boolean value
+   */
+  public static isInIPv4AddressRange(ipAddress: string, range: IPv4AddressRange): boolean {
+    if (this.isValidIPv4Address(ipAddress) && this.isValidIPv4Address(range.start)
+      && this.isValidIPv4Address(range.end)) {
+      const testComponents = ipAddress.split('.');
+      const lowComponents = range.start.split('.');
+      const highComponents = range.end.split('.');
+      for (let i = 0; i < 4; i++) {
+        if (!this.inIPv4AddressRange(lowComponents[i], highComponents[i], testComponents[i])) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Takes a set of IP addresses and a set of IP ranges, and indicates whether
+   * any of the IP addresses belong to any of the ranges.
+   * @param {string[]} ips: array of IP strings
+   * @param {IPv4AddressRange[]} ranges: array of range objects
+   * @return {boolean} boolean value
+   */
+  public static isAnyIPv4AddressInAnyIPv4AddressRange(ips: string[], ranges: Array<IPv4AddressRange>): boolean {
```

Comment:
I notice a common pattern throughout all methods. i.e. we're constantly dealing with the question of whether the input IP addresses are valid. It would be better for us to create an IPv4Address type that each of these methods could take as input (vs. a free-form string). You could define the IPv4Address type such that it can only be constructed with a valid IPv4 address string. By doing that, you would no longer need to handle the case where the input IP address(es) are invalid within all of these utilities.